core/src/main/java/org/apache/accumulo/core/summary/Gatherer.java

```diff
@@ -0,0 +1,513 @@
+
+    Predicate<String> hashingFileSelector = file -> Math.abs(Hashing.murmur3_32().hashString(file).asInt()) % modulus == remainder;
+
+    Map<String,Map<String,List<TRowRange>>> filesGBL = getFilesGroupedByLocation(hashingFileSelector);
+
+    SummaryCollection partitionSummaries = new SummaryCollection();
+
+    ExecutorService execSrv = Executors.newCachedThreadPool();
+    try {
+      TInfo tinfo = Tracer.traceInfo();
+      while (true) {
+        Set<String> failedFiles = new HashSet<>();
+
+        List<Future<ProcessedFiles>> futures = new ArrayList<>();
+        for (Entry<String,Map<String,List<TRowRange>>> entry : filesGBL.entrySet()) {
+          HostAndPort location = HostAndPort.fromString(entry.getKey());
+          Map<String,List<TRowRange>> allFiles = entry.getValue();
+
+          Future<ProcessedFiles> future = execSrv.submit(new FilesProcessor(tinfo, location, allFiles));
+          futures.add(future);
+        }
+
+        for (Future<ProcessedFiles> future : futures) {
+          try {
+            ProcessedFiles pfiles = future.get();
+            if (pfiles.summaries.getTotalFiles() > 0) {
+              partitionSummaries.merge(pfiles.summaries, factory);
+            }
+
+            failedFiles.addAll(pfiles.failedFiles);
+          } catch (InterruptedException | ExecutionException e) {
+            throw new AccumuloException(e);
+          }
+        }
+
+        if (failedFiles.size() > 0) {
+          // get new locations just for failed files
+          Predicate<String> fileSelector = hashingFileSelector.and(file -> failedFiles.contains(file));
+          filesGBL = getFilesGroupedByLocation(fileSelector);
+          UtilWaitThread.sleep(250);
+        } else {
+          break;
+        }
+      }
+    } finally {
+      execSrv.shutdownNow();
+    }
+
+    return partitionSummaries;
+  }
+
+  public static interface FileSystemResolver {
+    FileSystem get(Path file);
+  }
+
+  /**
+   * This method will read summaries from a set of files.
+   */
+  public SummaryCollection processFiles(FileSystemResolver volMgr, Map<String,List<TRowRange>> files, BlockCache summaryCache, BlockCache indexCache,
+      ExecutorService srp) {
+    SummaryCollection fileSummaries = new SummaryCollection();
+
+    List<Future<SummaryCollection>> futures = new ArrayList<>();
+    for (Entry<String,List<TRowRange>> entry : files.entrySet()) {
+      Future<SummaryCollection> future = srp.submit(() -> {
+        List<RowRange> rrl = Lists.transform(entry.getValue(), RowRange::new);
+        return getSummaries(volMgr, entry.getKey(), rrl, summaryCache, indexCache);
+      });
+      futures.add(future);
+    }
+
+    for (Future<SummaryCollection> future : futures) {
+      try {
+        fileSummaries.merge(future.get(), factory);
+      } catch (ExecutionException | InterruptedException e) {
+        throw new RuntimeException(e);
```

Comment:
Should be a good citizen and re-set the interrupted state when catching `InterruptedExeption`.